### PR TITLE
[MIG][16.0] product_analytic_sale

### PR DESCRIPTION
**Pull request by rousseldenis**

### CODE REVIEW

**Comment by rousseldenis** [comment]
@alexis-via @bizzappdev

**Comment by rousseldenis** [comment]
@sbidoul

File: product_analytic_sale/models/sale_order_line.py
Comment (bizzappdev) [inline product_analytic_sale/models/sale_order_line.py]:
```suggestion
        lines_without_analytic_ids = self.env["sale.order.line"]
        for line in self:
            if line.product_id:
                ana_accounts = (
                    line.product_id.product_tmpl_id._get_product_analytic_accounts()
                )
                ana_account = ana_accounts["income"]
                if ana_account:
                    line.analytic_distribution = {ana_account.id: 100}
                else:
                    lines_without_analytic_ids |= line
            else:
                lines_without_analytic_ids |= line
        return super(
            SaleOrderLine, lines_without_analytic_ids
```

Comment (bizzappdev) [reply product_analytic_sale/models/sale_order_line.py]: IMO, this will save browsing again already browsed

Comment (rousseldenis) [reply product_analytic_sale/models/sale_order_line.py]: Indeed, before v16, union was quite not performant. It operates quite flawlessly now, even if last browse does not affect performances as ids already in cache.

Comment (rousseldenis) [reply product_analytic_sale/models/sale_order_line.py]: @bizzappdev Code updated

File: product_analytic_sale/models/sale_order_line.py
Comment (sbidoul) [inline product_analytic_sale/models/sale_order_line.py:29]: In #576 (product_analytic_purchase), `create` is also overridden. Why is it different here?

File: product_analytic_sale/models/sale_order_line.py
Comment (bizzappdev) [inline product_analytic_sale/models/sale_order_line.py:18]: Might be I am missing something but don't you there should be else condition for this also so recodrs will be covered in super

File: product_analytic_sale/models/sale_order_line.py
Comment (sbidoul) [inline product_analytic_sale/models/sale_order_line.py]: I have tested with standard Odoo:
- if there is an analytic account on product or category, it is added on the invoice, by the `product_analytic` module
- if an analytic distribution is set on the SO line, it has priority over product analytic

So I think this `_prepare_invoice_line` override is not necessary anymore.

File: product_analytic_sale/models/sale_order_line.py
Comment (rousseldenis) [reply product_analytic_sale/models/sale_order_line.py:18]: Totally right

File: product_analytic_sale/models/sale_order_line.py
Comment (rousseldenis) [reply product_analytic_sale/models/sale_order_line.py:29]: Inherited from old version of the module as the distribution was not computed. Not needed. To be changed there.

**Comment by rousseldenis** [comment]
@sbidoul @bizzappdev Last check, this one is :green_apple:  ?

**Review by bizzappdev — APPROVED** [review]

**Review by sbidoul — COMMENTED** [review]
LGTM. I'll wait for our integration tests for giving the approval here.

File: product_analytic_sale/readme/DESCRIPTION.rst
Comment (sbidoul) [inline product_analytic_sale/readme/DESCRIPTION.rst]: @rousseldenis this paragraph is obsolete?

File: product_analytic_sale/readme/DESCRIPTION.rst
Comment (sbidoul) [inline product_analytic_sale/readme/DESCRIPTION.rst]:
```suggestion
```

File: product_analytic_sale/readme/DESCRIPTION.rst
Comment (rousseldenis) [reply product_analytic_sale/readme/DESCRIPTION.rst]: Mmmmh, I will add the test :sweat_smile:

**Comment by PauMAVA** [comment]
@rousseldenis @sbidoul will this be merged soon? I would be interested in using it in production. Thanks!

**Comment by rousseldenis** [comment]
> @rousseldenis @sbidoul will this be merged soon? I would be interested in using it in production. Thanks!

Ususally, some reviews are required in OCA to be merged. Don't hesitate to do it :smiley:

**Review by PauMAVA — APPROVED** [review]

**Review by aritzolea — APPROVED** [review]
LGTM (technical review)

**Comment by sbidoul** [comment]
@rousseldenis if I'm not mistaken you still need to handle my comments above?

File: product_analytic_sale/readme/DESCRIPTION.rst
Comment (rousseldenis) [reply product_analytic_sale/readme/DESCRIPTION.rst]: Why ? This is a technical section for developers.

Comment (sbidoul) [reply product_analytic_sale/readme/DESCRIPTION.rst]: Because we must not add top level section titles in README fragments as this breaks the rendered document.

File: product_analytic_sale/readme/DESCRIPTION.rst
Comment (rousseldenis) [reply product_analytic_sale/readme/DESCRIPTION.rst]: @sbidoul In fact, this is covered by : https://github.com/OCA/account-analytic/pull/621/commits/5b203de8f6636c8bf7fafb28639955098c8fc473

**Comment by rousseldenis** [comment]
@sbidoul This is ready

File: product_analytic_sale/readme/DESCRIPTION.rst
Comment (rousseldenis) [reply product_analytic_sale/readme/DESCRIPTION.rst]: aaaah ok. Is this better ?

File: product_analytic_sale/models/sale_order_line.py
Comment (sbidoul) [inline product_analytic_sale/models/sale_order_line.py:13]: Why do we mention `expense` here?

File: product_analytic_sale/readme/DESCRIPTION.rst
Comment (sbidoul) [reply product_analytic_sale/readme/DESCRIPTION.rst]: Nope. If you really want a subsection, then use `~`, but really there is no reason to create a subsection just for one line.

File: product_analytic_sale/models/sale_order_line.py
Comment (houssine78) [reply product_analytic_sale/models/sale_order_line.py:13]: @rousseldenis did you get time to go ahead with the comment of Stéphane ?

**Review by houssine78 — APPROVED** [review]
LGTM

**Review by dreispt — APPROVED** [review]

**Pull request merged**
